ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -260,4 +452,359 @@ private void updateMapHelper(Map<String, Long> map, String key, Long value) {
+              processNewPut(currentScanResults, newValue);
+            }
+          }
+        }
+      } catch (Exception e) {
+        logger.error("Unexpected exception while running QueueProcessor", e);
+        metrics.blobStoreStatsErrorCount.inc();
+      }
+    }
+
+    void cancel() {
+      cancelled = true;
+    }
+  }
+
+  /**
+   * Runner that scans the entire index and populate a new {@link ScanResults} to start/extend the forecast coverage so
+   * new requests can be answered using the created {@link ScanResults} instead of another scan of the index.
+   */
+  private class IndexScanner implements Runnable {
+    private volatile boolean cancelled = false;
+    volatile ScanResults newScanResults;
+    long startTimeInMs;
+
+    @Override
+    public void run() {
+      try {
+        recentEntryQueueEnabled = false;
+        startTimeInMs = time.milliseconds();
+        newScanResults = new ScanResults(startTimeInMs, logSegmentForecastOffsetMs, bucketCount, bucketSpanInMs);
+        synchronized (scanLock) {
+          isScanning = true;
+        }
+        Offset firstCheckpoint = index.getCurrentEndOffset();
+        ConcurrentNavigableMap<Offset, IndexSegment> indexSegments = index.getIndexSegments();
+        Map<StoreKey, Long> deletedKeys = new HashMap<>();
+        // process the active index segment based on the firstCheckpoint in case index segment rolled over after the
+        // checkpoint is taken
+        Offset activeIndexSegmentStartOffset = indexSegments.floorKey(firstCheckpoint);
+        IndexSegment activeIndexSegment = indexSegments.get(activeIndexSegmentStartOffset);
+        long indexSegmentStartProcessTime = time.milliseconds();
+        List<IndexEntry> activeIndexEntries = getFilteredIndexEntries(activeIndexSegment, firstCheckpoint);
+        processIndexSegmentEntriesBackward(activeIndexSegment, activeIndexEntries, deletedKeys);
+        metrics.statsBucketingScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTime);
+        if (indexSegments.size() > 1) {
+          ConcurrentNavigableMap<Offset, IndexSegment> sealedIndexSegments =
+              indexSegments.subMap(index.getStartOffset(), activeIndexSegmentStartOffset);
+          for (IndexSegment indexSegment : sealedIndexSegments.descendingMap().values()) {
+            if (cancelled) {
+              return;
+            }
+            indexSegmentStartProcessTime = time.milliseconds();
+            List<IndexEntry> indexEntries = getIndexEntries(indexSegment);
+            processIndexSegmentEntriesBackward(indexSegment, indexEntries, deletedKeys);
+            metrics.statsBucketingScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTime);
+          }
+        }
+        recentEntryQueueEnabled = true;
+        Offset secondCheckpoint = index.getCurrentEndOffset();
+        if (secondCheckpoint.compareTo(firstCheckpoint) > 0) {
+          forwardScan(firstCheckpoint, secondCheckpoint);
```

Comment:
the 2nd checkpoint was taken using lets say snapshot 1 of indexSegments. forwardScan() might be working on a diff snapshot for IndexSegments. May be if you can get a snapshot and ensure to work on that, we can be consistent all over